xmodule/capa_block.py

```diff
@@ -1748,14 +1777,47 @@ def submit_problem(self, data, override_time=False):
             # self.lcp.context['attempt'] refers to the attempt number (1-based)
             self.lcp.context['attempt'] = self.attempts + 1
             correct_map = self.lcp.grade_answers(answers)
+
+            # print(f'\n\nCurrent Correct Map: {self.lcp.correct_map.get_dict()}\n\n')
+            # print(f'\n\nNew Correct Map: {correct_map.get_dict()}\n\n')
+
             # self.attempts refers to the number of attempts that did not
             # raise an error (0-based)
             self.attempts = self.attempts + 1
             self.lcp.done = True
+
+            # Updates some fields of the xblock:
+            # self.correct_map, self.input_state, self.student_answers, self.has_saved_answers
             self.set_state_from_lcp()
-            self.set_score(self.score_from_lcp(self.lcp))
+
+            print(f'\n\nCurrent Score: {self.score}\n\n')
+            print(f'\n\nNew Score: {self.score_from_lcp(self.lcp)}\n\n')
+
+            # -------------------------------------------------------------
+
+            new_score = self.score_from_lcp(self.lcp)
+
+            self.score_history.append(new_score.raw_earned)
+            print(f'\n\nScore History: {self.score_history}\n\n')
+
+            if self.grading_strategy == GradingStrategy.HIGHEST_ATTEMPT:
+                if new_score.raw_earned > self.score.raw_earned:
+                    self.set_score(new_score)
+            elif self.grading_strategy == GradingStrategy.LAST_ATTEMPT:
+                self.set_score(new_score)
+            elif self.grading_strategy == GradingStrategy.AVERAGE_ATTEMPT:
+                average_score = round(sum(self.score_history) / len(self.score_history), 2)
+                new_score = Score(raw_earned=average_score, raw_possible=self.max_score())
+                self.set_score(new_score)
```

Comment:
Can you refactor this like this:

```python
class GradingStrategy:
    name = None
    def __init__(self, name):
        self.name = name

    def calculate_score(self, new_grade, last_grade, historical_grade):
         raise NotImplementedError(...)
```

Then subclass it and refactor this code to simplify it:
```suggestion
            new_score = GRADING_STRATEGY_MAPPING[self.grading_strategy].calculate_score(...)
            self.set_score(new_score)
```